test/cpp/end2end/xds_end2end_test.cc

```diff
@@ -7523,6 +7518,28 @@ TEST_P(XdsEnabledServerTest, ListenerAddressMismatch) {
   EXPECT_TRUE(rpc_failed);
 }
 
+TEST_P(XdsEnabledServerTest, UseOriginalDstNotSupported) {
+  Listener listener;
+  listener.set_name(
+      absl::StrCat("grpc/server?xds.resource.listening_address=",
+                   ipv6_only_ ? "[::1]:" : "127.0.0.1:", backends_[0]->port()));
+  balancers_[0]->ads_service()->SetLdsResource(listener);
+  listener.mutable_address()->mutable_socket_address()->set_address(
+      ipv6_only_ ? "::1" : "127.0.0.1");
+  listener.mutable_address()->mutable_socket_address()->set_port_value(
+      backends_[0]->port());
+  listener.add_filter_chains()->add_filters()->mutable_typed_config()->PackFrom(
+      HttpConnectionManager());
+  listener.mutable_use_original_dst()->set_value(true);
+  balancers_[0]->ads_service()->SetLdsResource(listener);
+  CheckRpcSendFailure(1, RpcOptions().set_wait_for_ready(true));
```

Comment:
I don't think `wait_for_ready` is actually needed here.

Reply:
Because, the channel should be in an idle state rather than TRANSIENT_FAILURE so the RPC will not be fail unless the connection attempt finished even if it is not set `wait_for_ready`?

Reply:
Yes.  `wait_for_ready` only comes into play when the channel transitions into `TRANSIENT_FAILURE`, and that won't happen until the connection attempt has failed.  Using `wait_for_ready` prevents the call from failing when the channel goes into `TRANSIENT_FAILURE`, which means that the reason the call is ultimately failing is that it hits its deadline.  If you check the failure state of this RPC, I believe you'll see that it's `DEADLINE_EXCEEDED` instead of `UNAVAILABLE`, which is what we should be looking for in this test.

Hmm... That having been said, even if we don't use `wait_for_ready`, how do we know that the server has actually gotten the config by the time this RPC fails?  The RPC has a 1s timeout.  If the client fails to connect to the server before the server actually receives the config, then the RPC sent by the client will fail before we know for sure that the server would not start listening -- this part of the test would pass even though the server is not behaving correctly.

Of course, if that happens, then the following check (for the LDS resource being NACKed) will fail.  But do we know that that check is not flaky?  It seems like even if the server is behaving correctly, the test could fail some percentage of the time due to the server not yet having had time to NACK the resource by the time we check the status.

Maybe this test would be more reliable if we used `backends_[0]->notifier()->WaitOnServingStatusChange()` to check explicitly what the backend's status is?  Once we get the status, we know that the server has seen the config, and then we can check that the client's RPC fails and that the xDS server sees the NACK.

Actually, it seems like this same comment applies to basically all of the xDS-enabled gRPC server tests.  Maybe we should use the status notifier for all of these tests, now that we have it?

Reply:
>Hmm... That having been said, even if we don't use wait_for_ready, how do we know that the server has actually gotten the config by the time this RPC fails? The RPC has a 1s timeout. If the client fails to connect to the server before the server actually receives the config, then the RPC sent by the client will fail before we know for sure that the server would not start listening -- this part of the test would pass even though the server is not behaving correctly.

I'm guessing that it wasn't flaky because the 1s timeout was enough for the clients to connect.

>Maybe this test would be more reliable if we used backends_[0]->notifier()->WaitOnServingStatusChange() to check explicitly what the backend's status is? Once we get the status, we know that the server has seen the config, and then we can check that the client's RPC fails and that the xDS server sees the NACK.

We can't use the status notifier in cases where we expect a NACK, since the config fetcher does not see the error update for NACKs. It only sees the error update when there is a timeout/resource deletion/error in its own processing of the update.
I've updated the tests to loop over the response state so that the tests are more reliable.